testing/constraints-async-rest-3.7.txt

```diff
@@ -0,0 +1,17 @@
+# This constraints file is used to check that lower bounds
```

Comment:
Why is a new file needed for async rest? Why not simply test against the existing constraints file? Or am I missing something here?

Reply:
Having a separate file will allow us to test the minimum versions of the dependencies in the extra `async_rest`. If we add `aiohttp` to the existing constraints file then we won't have a configuration where `aiohttp` is not installed. For example, when we run `nox -s unit-*`, the `aiohttp` extra should not be installed. It should only be installed when we run `nox -s unit_w_async_rest_extra*`. 

It much easier to have this in a separate constraints file than hardcoded in `nox -s unit_w_async_rest_extra*`. We specifically want to test the minimum versions of dependencies in the extra and not the latest versions so simply having `pip install .[aiohttp]` in the nox session isn't sufficient.

Reply:
_(my response below copied from [a comment on gapic-generator PR #2195](https://github.com/googleapis/gapic-generator-python/pull/2195#discussion_r1776152166)) on the same issue)_

I'm not sure this is true. The `nox` documentation [says](https://nox.thea.codes/en/stable/config.html) "[session.install] invokes pip to install packages inside of the session’s virtualenv." We use `session.install` like [this](https://github.com/googleapis/gapic-generator-python/blob/ab19c5a09b6ba1222cae4a207c7e4a4da4f83e21/gapic/templates/noxfile.py.j2#L57), specifying `-c CONSTRAINTS_FILE`. The `-c` constraints files option for `pip` is [documented](https://pip.pypa.io/en/latest/user_guide/#constraints-files) as saying (emphasis added) "Constraints files are requirements files that only control which version of a requirement is installed, _not whether it is installed or not_" .

So in other words, removing `async_rest` from the constraints will not do what we want. It doesn't matter whether we have it there or not, as long as we make sure that we do not _install_ the extras. So I would argue it's better to *not* split the constraints files, but to have one test suite that installs the `async_rest` extra and one test suite that doesn't, both with the same constraints file. 

Does this make sense? LMK if I've misunderstood something.